### PR TITLE
librados,osd: rados hints for future access pattern

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/osd/ECTransaction.cc [hunk 1]
```diff
@@ -162,6 +162,12 @@ struct TransGenerator : public boost::static_visitor<void> {
 	 ++i) {
       assert(buffers.count(i->first));
       bufferlist &enc_bl = buffers[i->first];
+      i->second.set_alloc_hint(
+	get_coll_ct(i->first, op.oid),
+	ghobject_t(op.oid, ghobject_t::NO_GEN, i->first),
+	0, 0,
+	CEPH_OSD_ALLOC_HINT_FLAG_SEQUENTIAL_WRITE |
+	CEPH_OSD_ALLOC_HINT_FLAG_APPEND_ONLY);
```
Comment (athanatos) [inline src/osd/ECTransaction.cc]: This, I'm not sure about.  Do we want to have non-advisory versions?  Would bluestore (for example) be able to exploit objects created with non-advisory versions of these flags more?

**Comment by liewegas** [comment]
@jdurgin @athanatos okay, i'm ready to use these in bluestore now in order to inform whether to use compression and to pick a csum chunk size (bigger chunks => less metadata but inefficient random writes).

**Comment by liewegas** [comment]
@ifed01 @allensamuels

**Comment by liewegas** [comment]
retest this please

File: src/osd/ECTransaction.cc (hunk 1 above)
Comment (liewegas) [inline src/osd/ECTransaction.cc]: An enforcing option might also make sense, but not all of these hints make any sense as enforcing options, and even if we had an enforcing flag the hints could still be useful.  For example, I'm pretty sure we'll only get appends, but I still want to allow an overwrite in rare cases.

File: src/os/ObjectStore.h [hunk 2]
```diff
@@ -1738,6 +1746,7 @@ class ObjectStore {
         _op->oid = _get_object_id(oid);
         _op->expected_object_size = expected_object_size;
         _op->expected_write_size = expected_write_size;
+	_op->alloc_hint_flags = flags;
```
Comment (ifed01) [inline src/os/ObjectStore.h]: Don't know what's the purpose for use_tbl option but shouldn't we encode flags for that flow too?

File: src/include/rados.h [hunk 3]
```diff
@@ -464,6 +464,19 @@ enum {
 
 const char *ceph_osd_watch_op_name(int o);
 
+enum {
+	CEPH_OSD_ALLOC_HINT_FLAG_SEQUENTIAL_WRITE = 1,
+	CEPH_OSD_ALLOC_HINT_FLAG_RANDOM_WRITE = 2,
```
Comment (ifed01) [inline src/include/rados.h]: Any rationale to have specific flags for random write/read? Don't they duplicate corresponding sequential read/write flags set to 0?

Comment (liewegas) [inline src/include/rados.h]: The hint is a positive statement, so having no hints means that we don't know anything--not that the opposite is true.

File: src/os/ObjectStore.h (hunk 2 above)
Comment (liewegas) [inline src/os/ObjectStore.h]: use_tbl is legacy compat code that should be removed shortly

**Comment by athanatos** [comment]
lgtm

**Comment by jdurgin** [comment]
lgtm as well

**Pull request merged**